site/docs/latest/advanced/Encryption.md

```diff
@@ -0,0 +1,157 @@
+                return encKeys.get(keyName + ".public-key");
+            }
+
+            @Override
+            public EncryptionKeyInfo getPrivateKey(String keyName, Map<String, String> keyMeta) {
+                return encKeys.get(keyName + ".private-key");
+            }
+        }
+
+        PulsarClient pulsarClient = PulsarClient.create("http://localhost:8080");
+
+        ProducerConfiguration prodConf = new ProducerConfiguration();
+        prodConf.setCryptoKeyReader(
+                new EncKeyReader("myappkey", "public key value".getBytes(), "private key value".getBytes()));
+        prodConf.addEncryptionKey("myappkey");
+
+        Producer producer = pulsarClient.createProducer("persistent://my-property/use/my-ns/my-topic", prodConf);
+
+        for (int i = 0; i < 10; i++) {
+            producer.send("my-message".getBytes());
+        }
+
+        pulsarClient.close();
+```
+7. Sample Consumer Application:
+```java
+        class EncKeyReader implements CryptoKeyReader {
+
+            Map<String, EncryptionKeyInfo> encKeys = new HashMap<String, EncryptionKeyInfo>();
+
+            EncKeyReader(String keyName, byte[] pubKey, byte[] privKey) {
+
+                EncryptionKeyInfo pubKeyInfo = new EncryptionKeyInfo(pubKey, null);
+                encKeys.put(keyName + ".public-key", pubKeyInfo);
+
+                EncryptionKeyInfo privKeyInfo = new EncryptionKeyInfo(privKey, null);
+                encKeys.put(keyName + ".private-key", privKeyInfo);
+
+            }
+
+            @Override
+            public EncryptionKeyInfo getPublicKey(String keyName, Map<String, String> keyMeta) {
+                return encKeys.get(keyName + ".public-key");
+            }
+
+            @Override
+            public EncryptionKeyInfo getPrivateKey(String keyName, Map<String, String> keyMeta) {
+                return encKeys.get(keyName + ".private-key");
+            }
+        }
+
+        ConsumerConfiguration consConf = new ConsumerConfiguration();
+        consConf.setCryptoKeyReader(
+                new EncKeyReader("myappkey", "public key value".getBytes(), "private key value".getBytes()));
+
+        PulsarClient pulsarClient = PulsarClient.create("http://localhost:8080");
+
+        Consumer consumer = pulsarClient.subscribe("persistent://my-property/use/my-ns/my-topic", "my-subscriber-name", consConf);
+        Message msg = null;
+
+        for (int i = 0; i < 10; i++) {
+            msg = consumer.receive();
+            // do something
+            System.out.println("Received: " + new String(msg.getData()));
+        }
+
+        // Acknowledge the consumption of all messages at once
+        consumer.acknowledge(msg);
+        pulsarClient.close();
+```
+## Key rotation
+Pulsar generates new AES data key every 4 hours or after a certain number of messages are published. The asymmetric public key is automatically fetched by producer every 4 hours to retrieve the latest version.
```

Comment:
What do you mean by "fetched by producer". Where is the producer fetching the key from?

Reply:
The producer calls CryptoKeyReader::getPublicKey()  every 4 hours to get the key and caches it. Updated the doc. The interface can directly read the file/invoking kms API to get latest key.